File: image_viewer.rb
#!/usr/bin/env ruby
# image_viewer.rb
# encoding: UTF-8

require 'rmagick'
include Magick
require 'io/console'

COLORS = {
  reset: "\e[0m",
  bold: "\e[1m"
}

def rgb_to_ansi(r, g, b)
  "\e[38;2;#{r};#{g};#{b}m"
end

def get_image_files(path)
  return [path] if File.file?(path)
  if File.directory?(path)
    exts = %w[.jpg .jpeg .png .gif .bmp .tiff .webp]
    Dir.entries(path)
       .select { |f| exts.include?(File.extname(f).downcase) }
       .map { |f| File.join(path, f) }
       .sort
  else
    []
  end
end

def display_image(img, zoom, rotate, flip_h, flip_v, ascii_mode, term_w, term_h)
  system('clear') || system('cls')
  w, h = img.columns, img.rows
  max_w = term_w - 2
  max_h = term_h - 6
  scale = zoom == 1.0 ? [max_w.to_f / w, max_h.to_f / h, 1.0].min : zoom
  new_w = [(w * scale).to_i, 1].max
  new_h = [(h * scale).to_i, 1].max

  processed = img.scale(new_w, new_h)
  processed = processed.rotate(rotate) if rotate != 0
  processed = processed.flip! if flip_h
  processed = processed.flop! if flip_v

  puts "#{COLORS[:bold]}📷 Image Viewer | #{new_w}×#{new_h}#{COLORS[:reset]}"
  puts "Zoom: #{zoom.round(1)}x | Rotate: #{rotate}° | Flip: #{flip_h}/#{flip_v}"
  puts "Controls: ← → (nav)  +/- (zoom)  r (rotate)  h/v (flip)  f (mode)  q (quit)"

  if ascii_mode
    chars = ' .:-=+*#%@'
    processed.to_blob { self.format = 'GRAY' }.bytes.each_slice(new_w) do |row|
      puts row.map { |b| chars[(b.to_f / 255 * (chars.length-1)).to_i] }.join
    end
  else
    pixels = processed.get_pixels(0, 0, new_w, new_h)
    new_h.times do |y|
      line = ''
      new_w.times do |x|
        pixel = pixels[y * new_w + x]
        line += rgb_to_ansi(pixel.red / 257, pixel.green / 257, pixel.blue / 257) + '██'
      end
      puts line + COLORS[:reset]
    end
  end
end

def getch
  STDIN.getch
end

def main
  path = '.'
  zoom = 1.0
  rotate = 0
  ascii_mode = false
  convert_format = nil
  output_file = nil

  i = 0
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when '-z' then zoom = ARGV[i+1].to_f; i += 1
    when '-r' then rotate = ARGV[i+1].to_i; i += 1
    when '-a' then ascii_mode = true
    when '-c' then convert_format = ARGV[i+1]; i += 1
    when '-o' then output_file = ARGV[i+1]; i += 1
    when '-h', '--help'
      puts "Usage: ruby image_viewer.rb <path> [-z zoom] [-r rotate] [-a] [-c format] [-o output]"
      return
    else path = arg
    end
    i += 1
  end

  files = get_image_files(path)
  if files.empty?
    puts "Нет изображений."
    return 1
  end

  idx = 0
  flip_h, flip_v = false, false
  img = Image.read(files[idx]).first

  if convert_format && output_file
    img.write(output_file)
    puts "✅ Изображение сохранено в #{output_file}"
    return
  end

  term_w = IO.console.winsize[1]
  term_h = IO.console.winsize[0]

  loop do
    display_image(img, zoom, rotate, flip_h, flip_v, ascii_mode, term_w, term_h)
    key = getch
    case key
    when 'q' then break
    when '+' then zoom = [3.0, zoom + 0.2].min
    when '-' then zoom = [0.2, zoom - 0.2].max
    when 'r' then rotate = (rotate + 90) % 360
    when 'h' then flip_h = !flip_h
    when 'v' then flip_v = !flip_v
    when 'f' then ascii_mode = !ascii_mode
    when "\e"
      seq = STDIN.read_nonblock(3) rescue nil
      if seq == '[D' && idx > 0
        idx -= 1
        img = Image.read(files[idx]).first
      elsif seq == '[C' && idx < files.size-1
        idx += 1
        img = Image.read(files[idx]).first
      end
    end
  end
end

main if __FILE__ == $0

File: image_viewer.js
// image_viewer.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const sharp = require('sharp');
const { execSync } = require('child_process');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    bold: '\x1b[1m'
};

function rgbToAnsi(r, g, b) {
    return `\x1b[38;2;${r};${g};${b}m`;
}

function getImageFiles(dir) {
    if (fs.statSync(dir).isFile()) return [dir];
    const exts = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.tiff', '.webp'];
    const files = fs.readdirSync(dir)
        .filter(f => exts.includes(path.extname(f).toLowerCase()))
        .map(f => path.join(dir, f))
        .sort();
    return files;
}

async function displayImage(imgPath, zoom, rotate, flipH, flipV, asciiMode, termW, termH) {
    console.clear();
    let img = sharp(imgPath);
    const meta = await img.metadata();
    let w = meta.width, h = meta.height;

    // Поворот
    if (rotate === 90) { img = img.rotate(90); [w, h] = [h, w]; }
    else if (rotate === 180) { img = img.rotate(180); }
    else if (rotate === 270) { img = img.rotate(-90); [w, h] = [h, w]; }
    // Флип
    if (flipH) img = img.flop();
    if (flipV) img = img.flip();

    const maxW = termW - 2;
    const maxH = termH - 6;
    let scale = zoom;
    if (scale === 1.0) {
        const sx = maxW / w;
        const sy = maxH / h;
        scale = Math.min(sx, sy, 1);
    }
    const newW = Math.max(1, Math.round(w * scale));
    const newH = Math.max(1, Math.round(h * scale));
    img = img.resize(newW, newH, { kernel: sharp.kernel.lanczos2 });

    const buffer = await img.raw().toBuffer({ resolveWithObject: true });
    const data = buffer.data;
    const sw = buffer.info.width;
    const sh = buffer.info.height;

    console.log(`${COLORS.bold}📷 Image Viewer | ${sw}×${sh}${COLORS.reset}`);
    console.log(`Zoom: ${zoom.toFixed(1)}x | Rotate: ${rotate}° | Flip: ${flipH}/${flipV}`);
    console.log('Controls: ← → (nav)  +/- (zoom)  r (rotate)  h/v (flip)  f (mode)  q (quit)');

    if (asciiMode) {
        const chars = ' .:-=+*#%@';
        for (let y = 0; y < sh; y++) {
            let line = '';
            for (let x = 0; x < sw; x++) {
                const idx = (y * sw + x) * 3;
                const r = data[idx], g = data[idx+1], b = data[idx+2];
                const gray = 0.299 * r + 0.587 * g + 0.114 * b;
                const ci = Math.floor((gray / 255) * (chars.length - 1));
                line += chars[ci];
            }
            console.log(line);
        }
    } else {
        for (let y = 0; y < sh; y++) {
            let line = '';
            for (let x = 0; x < sw; x++) {
                const idx = (y * sw + x) * 3;
                const r = data[idx], g = data[idx+1], b = data[idx+2];
                line += rgbToAnsi(r, g, b) + '██';
            }
            line += COLORS.reset;
            console.log(line);
        }
    }
}

function getKey() {
    return new Promise(resolve => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: true
        });
        process.stdin.setRawMode(true);
        process.stdin.once('data', (data) => {
            process.stdin.setRawMode(false);
            rl.close();
            resolve(data.toString());
        });
    });
}

async function main() {
    const args = process.argv.slice(2);
    let pathArg = '.', zoom = 1.0, rotate = 0, asciiMode = false;
    let convertFormat = '', outputFile = '';
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-z' && i+1 < args.length) zoom = parseFloat(args[++i]);
        else if (args[i] === '-r' && i+1 < args.length) rotate = parseInt(args[++i]);
        else if (args[i] === '-a') asciiMode = true;
        else if (args[i] === '-c' && i+1 < args.length) convertFormat = args[++i];
        else if (args[i] === '-o' && i+1 < args.length) outputFile = args[++i];
        else if (args[i] === '-h' || args[i] === '--help') {
            console.log('Usage: node image_viewer.js <path> [-z zoom] [-r rotate] [-a] [-c format] [-o output]');
            process.exit(0);
        } else pathArg = args[i];
    }

    const files = getImageFiles(pathArg);
    if (files.length === 0) { console.log('Нет изображений.'); process.exit(1); }
    let idx = 0, flipH = false, flipV = false;

    if (convertFormat && outputFile) {
        await sharp(files[idx]).toFile(outputFile);
        console.log(`✅ Изображение сохранено в ${outputFile}`);
        return;
    }

    const termW = process.stdout.columns || 80;
    const termH = process.stdout.rows || 24;

    while (true) {
        await displayImage(files[idx], zoom, rotate, flipH, flipV, asciiMode, termW, termH);
        const key = await getKey();
        if (key === 'q') break;
        else if (key === '+') zoom = Math.min(3.0, zoom + 0.2);
        else if (key === '-') zoom = Math.max(0.2, zoom - 0.2);
        else if (key === 'r') rotate = (rotate + 90) % 360;
        else if (key === 'h') flipH = !flipH;
        else if (key === 'v') flipV = !flipV;
        else if (key === 'f') asciiMode = !asciiMode;
        else if (key === '\x1b') {
            // стрелка
            const seq = await getKey();
            if (seq === '[D' && idx > 0) idx--;
            else if (seq === '[C' && idx < files.length-1) idx++;
        }
    }
}

main().catch(console.error);
